`NestedScrollView`'s outer scrollable jumping with `BouncingScrollPhysics` due to `double` precision errors

Michal-MK opened the pull request:
This PR fixes scrolling issues with `NestedScrollView` using the `BouncingScrollPhysics`. In one of the steps of the calculation, we can reach a state where the position of the inner scrollable is set to a `double` value that falls within `precisionErrorTolerance` of `0` but we were using `==` with `0` rather than checking for a precision. My posts in the linked issue show the current behavior, and how I reached to conclusion (the code in this PR). This PR only addresses the "jumping" of the outer scrollable.  

Fixes #136199

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

I have not finished a test for this since I have never done so and therefore have 0 experience writing tests in Flutter, so any help there would be appreciated. I am also not sure how to test double precision errors in general. I did run all the nested_scroll_view_tests.dart locally and there are no failures.

Michal-MK committed:
Drop double precision errors that result in values < 1e-10 causing the outer scrollable to jump.

Michal-MK committed:
Add missing type annotation

goderbauer commented:
@Michal-MK thanks for the contribution. Looks like this PR is failing some checks. Can you please take a look and fix those up? Thanks!

Michal-MK commented:
@goderbauer I can decode the Linux Analyze check, which is clear to me, I have a TODO in tests, but the issue is I do not know how to write tests for this fix. For the other 6 failing checks I have no idea what the output is trying to tell me. I looked at the "undestading a LUCI build failure", but that to me appears to describe something else.

```
â��05:52:02â�� Test failed.
â��â��â��â��â��â��â��â��â��â��â��â��â��â��â��â��â��â��â��â��â��â��â��â��â��â��â��â��â��â��â��â��â��â��â��â��â��â��â��â��â��â��â��â��â��â��â��â��â��â��â��â��â��â��â��â��
For your convenience, the error messages reported above are repeated here:
  ð���  ð���  
```

Michal-MK commented:
Apart from tree-status all tests should pass, I'll wait and see if a test for this will be needed/asked for.

Piinks reviewed (commented):
Thanks for the contribution @Michal-MK!
This will need a test to verify it works as expected, and to prevent us from accidentally undoing this in the future. :)

Michal-MK committed:
Replace another == 0 for double comparison
- added tests to check for flickering between 0/ less than 1e^(-10+) for outer scrollable (likely not intended as no tests were covering that)

Michal-MK commented:
Okay so I tried something, the issue is that I do not fully understand the test as I just while(true)'d until a set of inputs caused the `if` to be hit with the faulty values. Then I searched for a place where the wrong value would appear, and it appeared in the outer scrollable's controller.

I replaced another `== 0` check that may render the first one obsolete, and added a test that will fail if the outer scroll controller's listener is called. Since that is the place I found manifests the precision error to the user.

<img width="779" alt="image" src="https://github.com/flutter/flutter/assets/12874766/89a03121-a407-4fc5-9a41-6f33547157b9">

This is the output before the fix.
After the fix it does not trigger the outer controller's callback:

<img width="714" alt="image" src="https://github.com/flutter/flutter/assets/12874766/7ab0011b-76f2-4a40-b821-73d339b15050">

All other tests in NestedScrollView pass.

Piinks commented on packages/flutter/test/widgets/nested_scroll_view_test.dart:
```diff
@@ -212,6 +212,65 @@ void main() {
     expect(context.clipBehavior, equals(Clip.antiAlias));
   });
 
+  testWidgets('NestedScrollView always scrolls outer scrollable first, (double precision errors, #136199)', (WidgetTester tester) async {
+    final Key innerKey = UniqueKey();
```
```suggestion
  testWidgetsWithLeakTracking('NestedScrollView always scrolls outer scrollable first', (WidgetTester tester) async {
    // Regression test for https://github.com/flutter/flutter/issues/136199
    final Key innerKey = UniqueKey();
```

Piinks commented on packages/flutter/test/widgets/nested_scroll_view_test.dart:
```diff
@@ -212,6 +212,65 @@ void main() {
+    await tester.pumpWidget(build());
+
+    outerController.addListener(() {
+      assert(false); // We do not want this to be called.
```
Why? The outer should scroll before the inner, why would the outer not scroll?

Michal-MK replied:
The way I understand the test is that we have:
`tester.drag(find.byKey(innerKey), const Offset(0, 2000));`

Which means dragging the finger down the screen (I believe). So we are immediately overscrolling and triggering the dampening effect of the spring. The total height is 300 + 64 + 800 from the individual components. 1164 < 2000. So we have technically 836 units to work with.

`tester.fling(find.byKey(innerKey), const Offset(0, -600), 2000);`

After that we immediately fling the finger up the screen. My understanding is that the total distance will be 600 units up and the 2000 just affects how quickly the 600 will happen.

Which equals 836 - 600 = 236. So we still have 236 units to spare before the outer scrollable should start scrolling. What does not fly is the spring does not actually "stretch" to 836 visually, but I have no idea how that is implemented. If this description makes sense I will add it to the test itself, but there is a possibility it is nonsense and I just do not understand what I wrote ;).

Piinks replied:
Hmm, it may be easier to understand if you add checks on what the actual position is before and after each gesture. What is the value of position.pixels for the outer and inner each time? For the nested scroll view in general, the inner should not scroll before the outer has fully scrolled to its end.

If we are dragging down first, overscrolling the outer, then flinging up, I would expect the outer to consume some of the distance flinging up before the inner receives any.

Piinks replied:
> the outer is not being overscrolled

Ah right I see. The inner is using bouncing scroll physics. Yup. So the inner overscrolls at the top, but then when the test flings the inner up, I expect the outer to consume some of the distance before the inner does.

> Since we are not supposed to assign a scroll controller to the inner scrollview I do not have output from that.

The inner scroll controller is accessible through the inner key you have created in the test.

Can you add checks that validate the inner and outer positions at each step to the test?

Michal-MK replied:
Alright, I added more asserts to the test, I did not find a way to obtain the controller from the previous setup so I just borrowed someone else's solution :). The outer scrollable does not move, it may be surprising but it's true (according to the `expect()` and the `.addListener()`)

Piinks reviewed (commented):
Hey @Michal-MK welcome. Thanks for contributing!

Michal-MK committed:
Update packages/flutter/test/widgets/nested_scroll_view_test.dart

Co-authored-by: Kate Lovett <katelovett@google.com>

Piinks reviewed (commented):
@Michal-MK thanks for the update. It looks like there are a lot of failing tests currently, can you take a look?

Michal-MK commented:
Tests are failing randomly, as noted here: https://github.com/flutter/flutter/pull/138319#issuecomment-1813239714. No idea how to decode the output. I just merged current master into it and it worked again. That was the case even before I added my test so i do not know what is causing it.

Michal-MK committed:
Made the test more thorough

Michal-MK commented:
So it seems that Linux web_canvaskit_tests_4 and Linux web_tests_0 are now failing consistently, but their output is still illegible: â��â��â��â��â��â��â��â��â�

Michal-MK committed:
Cleanup and rework the test since Linux web_canvaskit_tests_4 and Linux web_tests_0 scroll differently than other platforms so constant values are not usable.

Michal-MK commented:
After spending more time with it I did find the tiny "show more details" button. There I finally got the actual output. Hmm why is GitHub's so broken then...? I modified the test to hopefully pass the two failing cases.

Michal-MK committed:
Add space after control flow statement

Piinks commented on packages/flutter/test/widgets/nested_scroll_view_test.dart:
```diff
@@ -212,6 +212,110 @@ void main() {
+
+    await tester.drag(find.byKey(innerKey), const Offset(0, 2000)); // Over-scroll the inner Scrollable to the bottom
+
+    const double endPosition = -1974.0862087158384;
```
Rather than this complicated value, you could use the `greaterThan`, `moreOrLessEquals`, etc tools that `flutter_test` provides.

Michal-MK replied:
Since. we are essentially testing a double precision error, I included the exact value the controller had after the drag, but I can change it.

Michal-MK replied:
On a second thought I like how the math adds up, because later in the code I do: endValue + 600 which would then not work... If I used `moreOrLessEquals` the delta would have to be really big. I could do `moreOrLessEquals(-2000, delta: 100.0)` sicne 2000 is the drag parameter, but I do not like how vague that becomes.

Piinks replied:
Ok! We should probably add some context at least in a comment for future wonderers. Style guide has recommendations: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#avoid-mysterious-and-magical-numbers-that-lack-a-clear-derivation

Piinks commented on packages/flutter/test/widgets/nested_scroll_view_test.dart:
```diff
@@ -212,6 +212,110 @@ void main() {
+    expect(outer.offset, 0.0);
+    expect(inner.offset, endPosition + 600);
+
+    // Since there is a difference between scrolling on devices and the web absolute values are not checked...
```
I am not sure I follow here. What was the failure this is working around?

Michal-MK replied:
Before I had concrete values at each step, as seen in this [commit](https://github.com/flutter/flutter/pull/138319/commits/7dd10eac04067f5b53d4df5b7a0ef1bca846d74b), I then rewrote it since tests on Web were failing due to the controller reporting different values than seen on native platforms. I can remove the comment if that is intended.

Piinks commented on packages/flutter/test/widgets/nested_scroll_view_test.dart:
```diff
@@ -212,6 +212,110 @@ void main() {
+      currentOffset = inner.offset;
+    }
+
+    // Assert positions returned to/stayed at 0.0
+    expect(outer.offset, 0.0);
+    expect(inner.offset, 0.0);
```
If you end with a `await tester.pumpAndSettle()` we can see where it actually ends as the fling comes to a rest. Can you add that?

Michal-MK replied:
Will do.

Piinks commented on packages/flutter/test/widgets/nested_scroll_view_test.dart:
```diff
@@ -212,6 +212,110 @@ void main() {
+
+    // Assert positions after pump
+    expect(outer.offset, 0.0);
+    expect(inner.offset, endPosition + 600);
```
Thank you for adding these checks! I much better understand what is happening here. The fling is consuming only the overscrolled portion of the inner scrollable for the whole test. 👍

Michal-MK replied:
👍

Michal-MK committed:
Modify tests

Piinks reviewed (commented):
This change looks good, with added context in the test for that value. I will be OOO on vacation after today, so if you would not like to wait, another reviewer can be requested on the Discord. This change will need two approvals. Thanks!

Michal-MK committed:
Add comment to precise endPosition variable, why it was used

Piinks reviewed (approved):
Thanks for the update @Michal-MK, and happy new year! This LGTM!

Michal-MK commented:
Happy new year to you and all of the Flutter team. Awesome framework, glad I could contribute. :)

Michal-MK commented:
@Piinks  Is there something on my part that needs doing? The "Google testing" test is now running for more than 5 days, is that intended?

Piinks commented:
This still needs a second review, but I have rebased it to kick the Google testing shard. 👍 
Following up on the second review in https://discord.com/channels/608014603317936148/804482791517192273/1193960740856545290

Piinks commented:
Hmm, I was not able to rebase since there is a merge conflict. Can you rebase with the master branch @Michal-MK?

Michal-MK commented:
Struggling with the rebase, give me a minute :)

Michal-MK committed:
Remove testWidgetsWithLeakTracking usage

Michal-MK commented:
So, I finally won agains git rebase, unfortunately, the bot added some people for review, that was not intentional. @Piinks I do not see the Discord channel and I have not found any info on how to enable it on what is publicly available.

Piinks commented:
Thanks for rebasing! Looks like that channel is restricted to contributors with commit access (members of flutter-hackers): https://github.com/flutter/flutter/wiki/Contributor-access

Sorry for sharing something inaccessible. The channel is for seeking reviewers, so it should get picked up soon. 👍

goderbauer reviewed (approved):
secondary LGTM
